### PR TITLE
Implement SpringBoot/REST call metadata interceptor

### PR DESCRIPTION
**Pull request by tbak**

### CODE REVIEW

File: titus-common-server/src/main/java/com/netflix/titus/runtime/endpoint/metadata/spring/CallMetadataAuthentication.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+package com.netflix.titus.runtime.endpoint.metadata.spring;
+
+import java.util.Collection;
+
+import com.netflix.titus.api.model.callmetadata.CallMetadata;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.GrantedAuthority;
+
+/**
+ * A decorator for an original {@link Authentication} object, which adds resolved request {@link CallMetadata}.
```
Comment (corindwyer) [inline titus-common-server/src/main/java/com/netflix/titus/runtime/endpoint/metadata/spring/CallMetadataAuthentication.java:26]: what does `an original Authentication object` mean? Are you wanting to convey that it is a standard concept from the Spring framework?

File: titus-common-server/src/main/java/com/netflix/titus/runtime/endpoint/metadata/spring/SpringCallMetadataInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    private static final String DEBUG_QUERY_PARAM = "debug";
+
+    @Override
+    public boolean preHandle(HttpServletRequest httpServletRequest, HttpServletResponse httpServletResponse, Object handler) throws Exception {
+        String callReason = httpServletRequest.getHeader(CallMetadataHeaders.CALL_REASON_HEADER);
+        boolean debug;
+        if (httpServletRequest.getParameter(DEBUG_QUERY_PARAM) == null) {
+            debug = Boolean.parseBoolean(httpServletRequest.getHeader(CallMetadataHeaders.DEBUG_HEADER));
+        } else {
+            debug = Boolean.parseBoolean(httpServletRequest.getParameter(DEBUG_QUERY_PARAM));
```
Comment (corindwyer) [inline titus-common-server/src/main/java/com/netflix/titus/runtime/endpoint/metadata/spring/SpringCallMetadataInterceptor.java:49]: could reuse `httpServletRequest.getParameter(DEBUG_QUERY_PARAM)`

File: titus-common-server/src/main/java/com/netflix/titus/runtime/endpoint/metadata/spring/SpringCallMetadataInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+                .withContext(getContextFromServletRequest(httpServletRequest))
+                .build();
+    }
+
+    protected Map<String, String> getContextFromServletRequest(HttpServletRequest httpServletRequest) {
+        Map<String, String> context = new HashMap<>();
+
+        context.put(CallMetadataHeaders.DIRECT_CALLER_CONTEXT_SERVICE_NAME, httpServletRequest.getRequestURI());
+        context.put(CallMetadataHeaders.DIRECT_CALLER_CONTEXT_SERVICE_METHOD, httpServletRequest.getMethod());
+        context.put(CallMetadataHeaders.DIRECT_CALLER_CONTEXT_TRANSPORT_TYPE, "HTTP");
```
Comment (corindwyer) [inline titus-common-server/src/main/java/com/netflix/titus/runtime/endpoint/metadata/spring/SpringCallMetadataInterceptor.java:100]: could use `httpServletRequest.isSecure()` to switch between `HTTP` and `HTTPS` here.

File: titus-common-server/src/main/java/com/netflix/titus/runtime/endpoint/metadata/spring/CallMetadataAuthentication.java (hunk 1 above)
Comment (tbak) [reply titus-common-server/src/main/java/com/netflix/titus/runtime/endpoint/metadata/spring/CallMetadataAuthentication.java:26]: I am referring here to the authenticated principal object that we are wrapping. `Authentication` interface is part of Spring security API.

File: titus-common-server/src/main/java/com/netflix/titus/runtime/endpoint/metadata/spring/SpringCallMetadataInterceptor.java (hunk 3 above)
Comment (tbak) [reply titus-common-server/src/main/java/com/netflix/titus/runtime/endpoint/metadata/spring/SpringCallMetadataInterceptor.java:100]: Transport type is HTTP or GRPC. Plain/TLS distinction has separate attribute. This code is copy-pasted from the previous (jersey) implementation. I want to keep it the same for compatibility reason.

**Pull request merged**